### PR TITLE
OSPF instances and missing mandatory fields fix attempt

### LINKED ISSUE #15701: Ospf module fails with Dell N2048 if OSPF is not running
### The problem

OSPF poller module crashes due to empty `ospfAreaBdrRtrStatus` if executed against Dell N2048P that isn't running OSPF. 

Actual crash happens when inserting a row into `ospf_instance` table. If OSPF isn't running OSPF-MIB::ospfGeneralGroup is missing many of the mandatory fields, like `ospfAreaBdrRtrStatus` and `ospfExternLsaCount`:

```OSPF-MIB::ospfRouterId.0 = IpAddress: 0.0.0.0
OSPF-MIB::ospfAdminStat.0 = INTEGER: disabled(2)
OSPF-MIB::ospfVersionNumber.0 = INTEGER: version2(2)
OSPF-MIB::ospfASBdrRtrStatus.0 = INTEGER: false(2)
OSPF-MIB::ospfTOSSupport.0 = INTEGER: false(2)
OSPF-MIB::ospfExtLsdbLimit.0 = INTEGER: -1
OSPF-MIB::ospfMulticastExtensions.0 = INTEGER: 0
OSPF-MIB::ospfExitOverflowInterval.0 = INTEGER: 0
OSPF-MIB::ospfDemandExtensions.0 = INTEGER: false(2)
```

### Output of ./validate.php

```txt
===========================================
Component | Version
--------- | -------
LibreNMS  | 24.1.0 (2024-01-07T17:49:52+02:00)
DB Schema | 2023_12_15_105529_access_points_nummonbssid_integer (276)
PHP       | 8.2.7
Python    | 3.11.2
Database  | MariaDB 10.11.4-MariaDB-1~deb12u1
RRDTool   | 1.7.2
SNMP      | 5.9.3
===========================================

[OK]    Composer Version: 2.6.6
[OK]    Dependencies up-to-date.
[OK]    Database connection successful
[OK]    Database Schema is current
[OK]    SQL Server meets minimum requirements
[OK]    lower_case_table_names is enabled
[OK]    MySQL engine is optimal
[OK]    Database and column collations are correct
[OK]    Database schema correct
[OK]    MySQl and PHP time match
[OK]    Active pollers found
[OK]    Dispatcher Service not detected
[OK]    Locks are functional
[OK]    Python poller wrapper is polling
[OK]    Redis is unavailable
[OK]    rrd_dir is writable
[OK]    rrdtool version ok
```

### What was the last working version of LibreNMS?

_No response_

### Anything in the logs that might be useful for us?

```txt
Error polling ospf module for 192.168.40.2. PDOException: SQLSTATE[HY000]: General error: 1364 Field 'ospfAreaBdrRtrStatus' doesn't have a default value in /opt/librenms/vendor/laravel/framework/src/Illuminate/>
Stack trace:
#0 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Connection.php(580): PDOStatement->execute()
#1 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Connection.php(788): Illuminate\Database\Connection->Illuminate\Database\{closure}()
#2 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Connection.php(755): Illuminate\Database\Connection->runQueryCallback()
#3 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Connection.php(569): Illuminate\Database\Connection->run()
#4 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Connection.php(533): Illuminate\Database\Connection->statement()
#5 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Query/Processors/Processor.php(32): Illuminate\Database\Connection->insert()
#6 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(3383): Illuminate\Database\Query\Processors\Processor->processInsertGetId()
#7 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(1925): Illuminate\Database\Query\Builder->insertGetId()
#8 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(1333): Illuminate\Database\Eloquent\Builder->__call()
#9 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(1298): Illuminate\Database\Eloquent\Model->insertAndSetId()
#10 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(1137): Illuminate\Database\Eloquent\Model->performInsert()
#11 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(584): Illuminate\Database\Eloquent\Model->save()
#12 /opt/librenms/vendor/laravel/framework/src/Illuminate/Support/helpers.php(307): Illuminate\Database\Eloquent\Builder->Illuminate\Database\Eloquent\{closure}()
#13 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(583): tap()
#14 /opt/librenms/vendor/laravel/framework/src/Illuminate/Support/Traits/ForwardsCalls.php(23): Illuminate\Database\Eloquent\Builder->updateOrCreate()
#15 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(2333): Illuminate\Database\Eloquent\Model->forwardCallTo()
#16 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(2345): Illuminate\Database\Eloquent\Model->__call()
#17 /opt/librenms/LibreNMS/Modules/Ospf.php(91): Illuminate\Database\Eloquent\Model::__callStatic()
#18 /opt/librenms/LibreNMS/Poller.php(176): LibreNMS\Modules\Ospf->poll()
#19 /opt/librenms/LibreNMS/Poller.php(103): LibreNMS\Poller->pollModules()
#20 /opt/librenms/app/Console/Commands/DevicePoll.php(44): LibreNMS\Poller->poll()
#21 /opt/librenms/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): App\Console\Commands\DevicePoll->handle()
#22 /opt/librenms/vendor/laravel/framework/src/Illuminate/Container/Util.php(41): Illuminate\Container\BoundMethod::Illuminate\Container\{closure}()
#23 /opt/librenms/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(93): Illuminate\Container\Util::unwrapIfClosure()
#24 /opt/librenms/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(35): Illuminate\Container\BoundMethod::callBoundMethod()
#25 /opt/librenms/vendor/laravel/framework/src/Illuminate/Container/Container.php(662): Illuminate\Container\BoundMethod::call()
#26 /opt/librenms/vendor/laravel/framework/src/Illuminate/Console/Command.php(208): Illuminate\Container\Container->call()
#27 /opt/librenms/vendor/symfony/console/Command/Command.php(326): Illuminate\Console\Command->execute()
#28 /opt/librenms/vendor/laravel/framework/src/Illuminate/Console/Command.php(177): Symfony\Component\Console\Command\Command->run()
#29 /opt/librenms/vendor/symfony/console/Application.php(1081): Illuminate\Console\Command->run()
#30 /opt/librenms/vendor/symfony/console/Application.php(320): Symfony\Component\Console\Application->doRunCommand()
#31 /opt/librenms/vendor/symfony/console/Application.php(174): Symfony\Component\Console\Application->doRun()
#32 /opt/librenms/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(201): Symfony\Component\Console\Application->run()
#33 /opt/librenms/app/Console/Kernel.php(66): Illuminate\Foundation\Console\Kernel->handle()
#34 /opt/librenms/lnms(38): App\Console\Kernel->handle()
#35 {main}
```

### PR DESCRIPTION
**Pull request by PipoCanaja**
We have a lot of columns that cannot be NULL, so we need to check that before trying to createOrUpdate

|           id          | int(10) unsigned | NO | PRI | NULL | auto_increment |
|:---------------------:|:----------------:|:--:|:---:|:----:|:--------------:|
| ospfRouterId          | varchar(32)      | NO |     | NULL |                |
| ospfAdminStat         | varchar(32)      | NO |     | NULL |                |
| ospfVersionNumber     | varchar(32)      | NO |     | NULL |                |
| ospfAreaBdrRtrStatus  | varchar(32)      | NO |     | NULL |                |
| ospfASBdrRtrStatus    | varchar(32)      | NO |     | NULL |                |
| ospfExternLsaCount    | int(10) unsigned | NO |     | NULL |                |
| ospfExternLsaCksumSum | int(11)          | NO |     | NULL |                |
| ospfTOSSupport        | varchar(32)      | NO |     | NULL |                |
| ospfOriginateNewLsas  | int(10) unsigned | NO |     | NULL |                |
| ospfRxNewLsas         | int(10) unsigned | NO |     | NULL |                |

fixes #15701 

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [x] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [x] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Comment by PipoCanaja** [comment]
something like:
```
foreach ($col in ['ospfRouterId', 'ospfRxNewLsas', 'ospfOriginateNewLsas', xxxxxxxxxx ]) {
    if (empty($ospf_entry[$col])) {
        continue; // skip invalid data
    }
}
```

File: LibreNMS/Modules/Ospf.php [hunk 1]
```diff
@@ -87,6 +87,11 @@ public function poll(OS $os, DataStorageInterface $datastore): void
                 if (empty($ospf_entry['ospfRouterId'])) {
                     continue; // skip invalid data
                 }
+                foreach (['ospfRxNewLsas', 'ospfOriginateNewLsas', 'ospfAreaBdrRtrStatus', 'ospfExternLsaCount', 'ospfTOSSupport', 'ospfExternLsaCksumSum', 'ospfExternLsaCount', 'ospfASBdrRtrStatus', 'ospfVersionNumber', 'ospfAdminStat'] as $column) {
```
Comment (vhuk) [inline LibreNMS/Modules/Ospf.php:90]: ospfExternLsaCount is twice on the list

**Review by vhuk — CHANGES_REQUESTED** [review]
duplicate ospfExternLsaCount, otherwise looks good to me.

File: LibreNMS/Modules/Ospf.php (hunk 1 above)
Comment (PipoCanaja) [reply LibreNMS/Modules/Ospf.php:90]: On my way to fix that. I suppose double check is not necessary :D Did you check it in your instance to confirm the issue is gone ?

File: LibreNMS/Modules/Ospf.php [hunk 2]
```diff
@@ -87,7 +87,7 @@ public function poll(OS $os, DataStorageInterface $datastore): void
                 if (empty($ospf_entry['ospfRouterId'])) {
                     continue; // skip invalid data
                 }
-                foreach (['ospfRxNewLsas', 'ospfOriginateNewLsas', 'ospfAreaBdrRtrStatus', 'ospfExternLsaCount', 'ospfTOSSupport', 'ospfExternLsaCksumSum', 'ospfExternLsaCount', 'ospfASBdrRtrStatus', 'ospfVersionNumber', 'ospfAdminStat'] as $column) {
+                foreach (['ospfRxNewLsas', 'ospfOriginateNewLsas', 'ospfAreaBdrRtrStatus', 'ospfTOSSupport', 'ospfExternLsaCksumSum', 'ospfExternLsaCount', 'ospfASBdrRtrStatus', 'ospfVersionNumber', 'ospfAdminStat'] as $column) {
                     if (! array_key_exists($column, $ospf_entry) || is_null($ospf_entry[$column])) {
                         continue; // This column must exist and not be null
```
Comment (vhuk) [inline LibreNMS/Modules/Ospf.php:92]: This would need to be `continue 2;`.

**Review by vhuk — CHANGES_REQUESTED** [review]

File: LibreNMS/Modules/Ospf.php (hunk 2 above)
Comment (PipoCanaja) [reply LibreNMS/Modules/Ospf.php]: Thanx for pointing it. fixed.

**Review by vhuk — APPROVED** [review]
Looks good to me. Thank you very much!

**Comment by PipoCanaja** [comment]
Great, then ready for review !

**Review by laf — APPROVED** [review]
LGTM

**Pull request merged**